code/_onclick/item_attack.dm

```diff
@@ -25,27 +25,46 @@ avoid code duplication. This includes items that may sometimes act as a standard
 		return
 	return
 
-// Called at the start of resolve_attackby(), before the actual attack.
-/obj/item/proc/pre_attack(atom/a, mob/user)
-	return
+/**
+ * Called at the start of resolve_attackby(), before the actual attack.
+ *
+ * Arguments:
+ * * atom/A - The atom about to be hit
+ * * mob/living/user - The mob doing the htting
+ * * params - click params such as alt/shift etc
+ *
+ * See: [/obj/item/proc/melee_attack_chain]
+ */
+
+/obj/item/proc/pre_attack(atom/A, mob/user, params) //do stuff before attackby!
+	if(SEND_SIGNAL(src, COMSIG_ITEM_PRE_ATTACK, A, user, params) & COMPONENT_CANCEL_ATTACK_CHAIN)
+		return TRUE
+	return FALSE //return TRUE to avoid calling attackby after this proc does stuff
 
 //I would prefer to rename this to attack(), but that would involve touching hundreds of files.
 /obj/item/proc/resolve_attackby(atom/A, mob/user, var/attack_modifier = 1, var/click_parameters)
-	pre_attack(A, user)
 	add_fingerprint(user)
+	. = pre_attack(A, user, click_parameters)
+	if(.)	// We're returning the value of pre_attack, in the event it has a special return rather than a boolean.
```

Comment:
Comment is misleading - this will return for any truthy value, not simply a special (assuming string/instance/numeric?) value.